samples/nrf9160/lwm2m_client/src/main.c

```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
 #define LWM2M_SECURITY_CERTIFICATE 2
 #define LWM2M_SECURITY_NO_SEC 3
 
+/* Client State Machine states*/
+enum lwm2m_client_state {
+	START,		/* Start Connection to a server*/
+	CONNECTING,	/* LwM2M engine is connecting to server */
+	BOOTSTRAP,	/* LwM2M engine is doing a bootstrap */
+	CONNECTED,	/* LwM2M Client connection establisment to server */
+	LTE_OFFLINE,	/* LTE offline and LwM2M engine should be suspended */
+	NETWORK_ERROR	/* Client networ error handling. Client stop and modem reset */
+};
+
+static enum lwm2m_client_state client_state = START;
+
 static uint8_t endpoint_name[ENDPOINT_NAME_LEN + 1];
 static uint8_t imei_buf[IMEI_LEN + sizeof("\r\nOK\r\n")];
 static struct lwm2m_ctx client = {0};
 static bool reconnect;
-static int reconnection_counter;
-static struct k_sem lwm2m_restart;
+static struct k_sem lwm2m_state_machine;
```

Comment:
Also we could use `K_SEM_DEFINE(state_mutex, 0, 1);` so we don't need to call `k_sem_init()`